src/mlpack/methods/ann/augmented/tasks/copy_impl.hpp

```diff
@@ -0,0 +1,107 @@
+/**
+ * @file copy_impl.hpp
+ * @author Konstantin Sidorov
+ *
+ * Implementation of CopyTask class
+ *
+ * mlpack is free software; you may redistribute it and/or modify it under the
+ * terms of the 3-clause BSD license.  You should have received a copy of the
+ * 3-clause BSD license along with mlpack.  If not, see
+ * http://www.opensource.org/licenses/BSD-3-Clause for more information.
+ */
+#ifndef MLPACK_METHODS_AUGMENTED_TASKS_COPY_IMPL_HPP
+#define MLPACK_METHODS_AUGMENTED_TASKS_COPY_IMPL_HPP
+
+// In case it hasn't been included yet.
+#include "copy.hpp"
+
+namespace mlpack {
+namespace ann /* Artificial Neural Network */ {
+namespace augmented /* Augmented neural network */ {
+namespace tasks /* Task utilities for augmented */ {
+
+CopyTask::CopyTask(const size_t maxLength,
+                   const size_t nRepeats,
+                   const bool addSeparator) :
+    maxLength(maxLength),
+    nRepeats(nRepeats),
+    addSeparator(addSeparator)
+{
+  if (maxLength <= 1) 
+  {
+    std::ostringstream oss;
+    oss << "CopyTask::CopyTask(): maximum sequence length ("
+        << maxLength << ") "
+        << "should be at least 2!"
+        << std::endl;
+    throw std::invalid_argument(oss.str());
+  }
+  if (nRepeats <= 0)
+  {
+    std::ostringstream oss;
+    oss << "CopyTask::CopyTask(): repetition count (" << nRepeats << ") "
+        << "is not positive!"
+        << std::endl;
+    throw std::invalid_argument(oss.str());
+  }
+  // Just storing task-specific parameters.
+}
+
+const void CopyTask::Generate(arma::field<arma::mat>& input,
+                              arma::field<arma::mat>& labels,
+                              const size_t batchSize,
+                              bool fixedLength)
+{
+  input = arma::field<arma::mat>(batchSize);
+  labels = arma::field<arma::mat>(batchSize);
+  size_t size = maxLength;
+  for (size_t i = 0; i < batchSize; ++i)
+  {
+    if (!fixedLength)
+    {
+      // Generate random uniform length from [2..maxLength].
+      size = mlpack::math::RandInt(2, maxLength+1);
```

Comment:
For this reason, this generator (if I'm not mistaken) samples uniformly from the space of nonempty binary sequences with length *less than or equal* to `maxLength` (`{0, 1}^1 + {0, 1}^2 + ... + {0, 1}^L`)